packages/doenetml-worker-rust/lib-doenetml-core/src/core/component_builder.rs

```diff
@@ -362,7 +376,60 @@ impl ComponentBuilder {
                         unrecognized_attributes: HashMap::new(),
                     },
                 );
-                if let Some(Source::Ref(ref_resolution)) = &elm.extending {
+
+                // A some of a component's attributes may specify that refs in those attributes should not be expanded.
+                // We mark the corresponding components.
+                // Because the component's attributes have not been processed yet, we need to look up the corresponding
+                // attributes from `node`
+                for preserve_ref_attr_name in component
+                    .get_preserve_ref_attribute_indices()
+                    .iter()
+                    .map(|idx| component.get_attribute_names()[*idx])
+                {
+                    let attr = elm
+                        .attributes
+                        .iter()
+                        .find(|a| a.name == preserve_ref_attr_name);
+                    if attr.is_none() {
+                        continue;
+                    }
+                    let attr = attr.unwrap();
+                    for child in &attr.children {
+                        match child {
+                            UntaggedContent::Ref(idx) => {
+                                self.component_preserve_refs[ComponentIdx::from(*idx)] = true;
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+
+                if self.component_preserve_refs[component.get_idx()] {
+                    // If we're here, we have special expanding behavior.
+                    // We want to expand _most_ components that are marked with `extending`, but not all.
+                    // Components that are children of an attribute that has the `preserve_refs` flag set
+                    // are protected and should be converted into a special `_ref` component instead of being linked
+                    // to their extend referent.
+                    if let Some(ref_source) = &elm.extending {
+                        let ref_resolution = match ref_source {
+                            Source::Ref(ref_resolution) => ref_resolution,
+                            _ => unreachable!("Encountered an extend of style <foo extend='$xxx' /> inside of an attribute. It should be syntactically impossible to parse such an element in an attribute."),
+                        };
+                        if ref_resolution.unresolved_path.is_some() {
+                            // XXX: this should result in the component being converted into an error, not a hard panic.
+                            panic!("Cannot preserve_refs if there is a remaining path part. Remaining: {:?}", ref_resolution.unresolved_path);
```

Comment:
Actually, we will eventually need the ability to preserve refs even with an remaining part. For example,
```
<point name="p">(3,2)</point>
<updateValue target="$p.x" newValue="$p.x+1" />
```
should increment the x-component of the point on each click of the button. And for `target` we want to preserve refs so we can find the target to update it.

Not that we need to do that now, but it should be a TODO if not an XXX.